### ISSUE TITLE
Make error responses and exceptions when proxying Google Drive more visible

### ISSUE THREAD

**Issue by seanh** [hypothesis/via#590]
When proxying Google Drive files in Python fails make it easier to find the original error response from Google Drive (or the exception from `requests` if it fails without us receiving a response) for debugging.

## Solution

There's a lot of details below but following some discussion in a Slack thread [we decided that](https://hypothes-is.slack.com/archives/C1MA4E9B9/p1630950177056500?thread_ts=1630946931.045000&cid=C1MA4E9B9):

* The error responses will be 417 Expectation Failed (as currently)
* The body will be some JSON containing the details of the error response or exception, not HTML as currently

Note that we also need to add the _body_ of any error response from Google into our JSON but there's [already a separate ticket for including response bodies](https://github.com/hypothesis/via/issues/584)

## Problem

The new Python-based Google Drive proxying makes error statuses and bodies from Google Drive (or exception details from `requests`) difficult to access which will make debugging Google Drive issues in production difficult for users, support staff and developers.

Specifically:

* We return a "417 Expectation Failed" status which can be seen in dev tools but isn't very helpful. The 417 code is probably fine. But could we replace the status message with something more suggestive? E.g. "417 Hypothesis got an error from Google Drive" (and it may vary depending on what exactly failed and where)
  * In the case where we do actually have an error response from Google Drive should we proxy the status of Google Drive's response directly? This might be what would make debugging easiest for everyone
* We return a complex HTML page for the body. This HTML page is where we put the error response from Google or the exception from `requests` (in cases where it fails without us even receiving a response). This error page is not rendered in the browser: it's returned to PDF.js's JavaScript code. If you look up the 417 response in the dev tools and look at the response body **the dev tools show the HTML source code** (not rendered) and the complex HTML code is hard to read in this form. Can we return a much smaller response that's easier to view in dev tools?
  * In the case where we do actually have an error response from Google Drive should we proxy the body of Google Drive's response directly? This might be what would make debugging easiest for everyone

Details:

There are at least four different places where the Google Drive proxying code can raise an exception:

1. If Google Drive returns an error response the call to `response.raise_for_status()` will raise
2. If there's a networking error or something like that then the call to `_session.get()` will raise and we won't even have an error response from Google, just an exception object from `requests`
3. The first call to `iter_content()` can raise. We haven't returned a response object to Pyramid yet so there's still time for us to return an error status and body
4. A subsequent call to `iter_content()` can raise. In this case it's too late for us to return anything useful in the response status or body. All we can do is make sure the exception gets reported to Sentry properly (which is a [separate issue](https://github.com/hypothesis/via/issues/585))

Each of the four cases in detail:

### 1. Exception from the `raise_for_status()` (with an error response from Google Drive)

TLDR: In order to see a 4xx or 5xx error status or body from Google the user has to:

1. Open the networking tab in the browser's dev tools (and possibly re-load the page, if dev tools weren't already open when they launched the assignment)
2. Find the 417 error in dev tools
3. Go to the response tab. You can now see the error body of our custom error page which includes the error status and body from Google but it's unreadable because you're seeing the HTML code of our error page in dev tools, it's not rendered
4. Right-click and select <kbd>Open in new tab</kbd>. You can now see our custom error page rendered. But this isn't the original error response, it's re-sending the request in a new tab and showing you the new response. This could be problematic with intermittent errors
5. Click to expand **Error Details**. This is where you'll find the error status and body from Google.

<details>

Diff to reproduce a `raise_for_status()` failure:

```diff
diff --git a/via/services/google_drive.py b/via/services/google_drive.py
index 50b1790..bf68881 100644
--- a/via/services/google_drive.py
+++ b/via/services/google_drive.py
@@ -85,7 +85,7 @@ class GoogleDriveAPI:
             }
         )
 
-        response = self._session.get(url=url, headers=headers, stream=True, timeout=10)
+        response = self._session.get(url="https://httpbin.org/status/403", headers=headers, stream=True, timeout=10)
         response.raise_for_status()
 
         yield from stream_bytes(response)
```

What the user sees:

![Screenshot from 2021-09-06 15-42-25](https://user-images.githubusercontent.com/22498/132233947-1ca70e04-4835-480d-924a-72451aee9976.png)

And after clicking the <kbd>More Information</kbd> button:

![Screenshot from 2021-09-06 15-43-48](https://user-images.githubusercontent.com/22498/132234066-defa6fa5-08bc-41c7-955c-310ff90eb424.png)

If you look in the **Network** tab in the browser's dev tools and find the `/google_drive/<ID>?url=<URL>` request with the 417 response and click on the **Response** tab then you'll see the actual error body that we sent. This is an error response body from our app, not the response body from Google. And the dev tools don't render it, they just show the HTML. Technically there could be more information about the error from Google Drive in here (we could include Google Drive's error body or an excerpt of it in our error page) but it won't be easy to find by reading the HTML:

![Screenshot from 2021-09-06 15-49-18](https://user-images.githubusercontent.com/22498/132234663-53c8f989-e87f-454c-b6ad-f6c6f35ee5af.png)

If you **right click** on this HTML code there is an <kbd>Open in new tab</kbd> button and that, finally, does actually show you the error body from our app rendered:

![Screenshot from 2021-09-06 15-53-10](https://user-images.githubusercontent.com/22498/132235001-320d6fba-4147-4c40-98c3-136bc84b82ea.png)

But:

1. This doesn't actually show you the original response. It **re-sends the request** in a new tab and shows you the new response that it gets. This may or may not be the same error response. For example if we're dealing with an intermittent error then when the browser re-sends the request it may succeed.

2. The user still has one more click to expand **Error Details** before they'll see the error body from Google

3. The <kbd>Try again</kbd> button is nonsensical in this context

</details>

### 2. Exception from `_session.get()` (no error response)

TLDR: All the user sees in PDF.js is the status code from us (409) not even the status message. This isn't very helpful but there may not be much we can do about it (it's PDF.js's fault). In dev tools the user will see the full status from us but it's just "409 Conflict". Could we use the status message to show something more useful? Our response **body** is also viewable in dev tools, it's a fancy error page that does actually details of what went wrong in the **Error Details** section (for example: `UpstreamServiceError: HTTPSConnectionPool(host='dsfsdfsdfsd', port=443): Max retries exceeded ... Failed to establish a new connection ... Temporary failure in name resolution`) but it's extremely unreadable because dev tools show you the HTML source they don't render it. The user can click <kbd>Open in new tab</kbd> to **re-send the request** and **if they get the same error** then they'll see our error page rendered, albeit with a non-sensical <kbd>Try again</kbd> button.

<details>

Diff to simulate this:

```diff
diff --git a/via/services/google_drive.py b/via/services/google_drive.py
index 80be077..0d94aea 100644
--- a/via/services/google_drive.py
+++ b/via/services/google_drive.py
@@ -82,7 +82,7 @@ class GoogleDriveAPI:
             }
         )
 
-        response = self._session.get(url=url, headers=headers, stream=True, timeout=10)
+        response = self._session.get(url="https://dsfsdfsdfsd/", headers=headers, stream=True, timeout=10)
         response.raise_for_status()
 
         yield from stream_bytes(response)
```

What the user sees:

The error message in PDF.js isn't very useful. But I'm not sure there's anything we can do about this: PDF.js shows only shows the status code from us, not the status message or anything else. We seem to use a particular code (409 Conflict) in this situation:

![Screenshot from 2021-09-06 16-26-51](https://user-images.githubusercontent.com/22498/132239001-96a9510b-0f60-4c86-8d16-fcae487ff9ff.png)

In developer tools the situation is the same as for `raise_for_status()` errors above: we return a custom error page and it's completely unreadable because the dev tools don't render the HTML:

![Screenshot from 2021-09-06 16-29-47](https://user-images.githubusercontent.com/22498/132239190-3a1bc786-0694-430d-aec4-57aceb7bc9f2.png)

Again you can right-click and click <kbd>Open in new tab</kbd> to **re-send the request** and render the new response. This will be problematic with intermittent errors. But if you do manage to reproduce the error then you'll see our custom error page:

![Screenshot from 2021-09-06 16-31-25](https://user-images.githubusercontent.com/22498/132239416-42c4db57-70aa-460d-b82e-56f98fb06e14.png)

Again the <kbd>Try again</kbd> button is non-sensical.

</details>

### 3. Exception from the first call to `iter_content()`

The code might get past both the call to `_session.get()` and the `response.raise_for_status()` but then get an exception from the first call to `iter_content()`. The first call to `iter_content()` is different than subsequent calls to `iter_content()` because we haven't returned a `200 OK` response object to Pyramid yet so it's not too late for us to send an error response.

<details>

The method we're calling here is [`requests.Response.iter_content()`](https://2.python-requests.org/en/master/api/#requests.Response.iter_content). There's no documentation of what exceptions that method might raise but looking at [the source code](https://2.python-requests.org/en/master/_modules/requests/models/#Response.iter_content) it can raise at least `ChunkedEncodingError`, `ContentDecodingError` and `ConnectionError`.

TLDR: It's the same issue as above. In this case it's a `requests` exception that we have to show rather than a Google Drive error response. But we make the exception details difficult to access in the same way: they're buried in a complex HTML error page that's hard to read in dev tools.

Diff to simulate this situation:

```diff
diff --git a/via/requests_tools/streaming.py b/via/requests_tools/streaming.py
index 7b9eb20..d28d444 100644
--- a/via/requests_tools/streaming.py
+++ b/via/requests_tools/streaming.py
@@ -11,6 +11,9 @@ def stream_bytes(response, min_chunk_size=64000):
     # or just about 9 bytes, so we'll do some smoothing for our callers so we
     # don't incur overhead with very short iterations of content
     for chunk in response.iter_content(chunk_size=min_chunk_size):
+        import pdb; pdb.set_trace()
+        from requests.exceptions import ChunkedEncodingError
+        raise ChunkedEncodingError(RuntimeError("Foobar"))
         buffer += chunk
         if len(buffer) >= min_chunk_size:
             yield buffer
```

What the user sees:

* 417 in PDF.js.
 
* "417 Expectation Failed" in dev tools. **Can we replace the status _message_ with something informative?**

* Source code of complex HTML error page in dev tools. **We need to replace the response with something much smaller that is readable in dev tools.**

* If they right-click in dev tools and select <kbd>Open in new tab</kbd> then it'll **re-send the request** in a new tab and **if the request gets the same error response again** the user will see our complex error page rendered and be able to read about the original exception in the **Error Details**. *But* the original exception's class name is not shown:

  ![Screenshot from 2021-09-06 17-00-14](https://user-images.githubusercontent.com/22498/132242325-31d8f84c-a90a-438b-b84f-600f2e0f4f27.png)

</details>

### 4. Exception from a subsequent call to `iter_content()`

The code might get past the call to `_session.get()` and the `response.raise_for_status()` **and** the first call to `iter_content()` but then get an exception from a subsequent call  to `iter_content()`. The difference here is that we've already returned a 200 OK response and the first chunk of a PDF body to Pyramid so it's too late for us to do anything useful with the response status or body. All we can do with these ones is [make sure that the exception gets reported to Sentry](https://github.com/hypothesis/via/issues/585).

I haven't figured out a way to simulate this situation yet...

**Comment by jon-betts** [comment]
It's worth mentioning here that the response isn't always 417, as stated as can be seen in some of the examples above. The actual behavior is:

* For any form of bad URL we receive we emit a 400 (this doesn't apply to Google Drive)
* For a 404 in Google Drive we emit a 404
* For `ConnectionError`, `Timeout`, `TooManyRedirects` or `SSLError` we emit an `UpstreamServiceError` - 409
* For all other errors we raise `UnhandledException` - 417

The 400 and 404 are obvious, but I'll talk a little about the others.

The description for 409 used in our `UpstreamServiceError` is:

> **The request could not be completed due to a conflict with the current state of the resource.** This code is only allowed in situations where it is expected that the user might be able to resolve the conflict and resubmit the request. 

It's a mis-use of the code to avoid using the correct code which is probably closer to a `502`: 

> The server, while acting as a gateway or proxy, received an invalid response from the upstream server it accessed in attempting to fulfill the request.

The description for 417 used in our `UnhandledException`is:

>The expectation given in an Expect request-header field (see section 14.20) could not be met by this server, or, **if the server is a proxy, the server has unambiguous evidence that the request could not be met by the next-hop server**.

This is true enough, and is an alternative to a 502 or 503.

**Comment by seanh** [comment]
I'm gonna close this in favour of two separate issues:

1. https://github.com/hypothesis/via/issues/584
2. https://github.com/hypothesis/via/issues/596
